text/XXX-annotate-unwind-rust.md

```diff
@@ -0,0 +1,191 @@
+        panic!("Oops, I should have used u32.");
+    }
+}
+```
+
+In a future (TBD) version of Rust, calling this function with an argument less
+than zero will cause the program to be aborted. This is the only way the Rust
+compiler can ensure that the function is safe, because there is no way for it
+to know whether or not the calling code supports the same implementation of
+stack-unwinding used for Rust.
+
+As a concrete example, Rust code marked `exern "C"` may be invoked from C code,
+but C code on Linux or BSD may be compiled without support for native
+stack-unwinding. This means that the runtime would lack the necessary metadata
+to properly perform the unwinding operation.
+
+However, there are cases in which a `panic` through an `extern` function can be
+used safely. For instance, it is possible to invoke `extern` Rust functions
+via Rust code built with the same toolchain, in which case it would be
+irrelevant to the unwinding operation that the `panic`ing function is an
+`extern` function:
+
+```rust
+fn main() {
+    let result = panic::catch_unwind(|| {
+        may_panic(-1);
+    }
+    assert!(result.is_err());
+}
+```
+
+In order to ensure that `may_panic` will not simply abort in a future version
+of Rust, it must be marked `#[unwind(Rust)]`. This annotation can only be used
+with an `unsafe` function, since the compiler is unable to make guarantees
+about the behavior of the caller.
+
+```rust
+#[unwind(Rust)]
+unsafe extern "C" fn may_panic(i: i32) {
+    if i < 0 {
+        panic!("Oops, I should have used u32.");
+    }
+}
+```
+
+**PLEASE NOTE:** Using this annotation **does not** provide any guarantees
+about the unwinding implementation. Therefore, using this feature to compile
+functions intended to be called by C code **does not make the behavior
+well-defined**; in particular, **the behavior may change in a future version of
+Rust.**
+
+The *only* well-defined behavior specified by this annotation is Rust-to-Rust
+unwinding.
+
+It is safe to call such a function from C or C++ code only if that code is
+guaranteed to provide the same unwinding implementation as the Rust compiler
+used to compile the function.
+
+Since the behavior may be subject to change without notice as the Rust compiler
+is updated, it is recommended that all projects that rely on unwinding from
+Rust code into C code lock the project's `rustc` version and only update it
+after ensuring that the behavior will remain correct.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Unwinding for functions marked `#[unwind(Rust)]` is performed as if the
+function were not marked `extern`. This is identical to the behavior of `rustc`
+for all versions prior to 1.35 except for 1.24.0.
+
+This annotation has no effect on functions not marked `extern`. It has no
+observable effect unless the marked function `panic`s (e.g. it has no
+observable effect when a function returns normally or enters an infinite loop).
```

Comment:
Note that the attribute does have an effect from the point-of-view of how the generated code can be optimized, even if the marked function never `panic!`s in practice.